Source/SwiftLintFramework/Rules/Style/ImplicitSelfRule.swift

```diff
@@ -0,0 +1,231 @@
+import Foundation
+import SourceKittenFramework
+
+public struct ImplicitSelfRule: OptInRule, ConfigurationProviderRule {
+    public var configuration = ImplicitSelfConfiguration(severity: .warning, initSelfUsage: .never)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "implicit_self",
+        name: "Implicit Self",
+        description: "Instance variables and methods should not be accessed with 'self.' unless necessary.",
+        kind: .style
+    )
+
+    public func validate(file: SwiftLintFile) -> [StyleViolation] {
+        let identifiers = file.syntaxMap.tokens.compactMap { token -> Identifier? in
+            guard let kind = token.kind, kind == .identifier else { return nil }
+
+            return Identifier(token: token, file: file)
+        }
+
+        let structure = Structure(dict: file.structureDictionary.value)
+
+        for identifier in identifiers {
+            structure.add(identifier: identifier)
+        }
+
+        return violations(in: structure, declInScope: [], file: file)
+    }
+
+    private func violations(in structure: Structure,
+                            declInScope: Set<String>,
+                            file: SwiftLintFile) -> [StyleViolation] {
+        var existingDeclInScope: Set<String> = declInScope
+        guard !structure.isClosure else { return [] }
+        if structure.isInit && configuration.initSelfUsage == .always { return [] }
+        let violations = structure.identifiers.compactMap { identifier -> StyleViolation? in
+            guard identifier.name != "init" else { return nil }
+
+            guard !identifier.isParamDecl && !identifier.isLocalDecl else {
+                existingDeclInScope.insert(identifier.name)
+                return nil
+            }
+
+            if configuration.initSelfUsage == .beforeInitCall &&
+                structure.isInit &&
```

Comment:
These two conditions are constant for all elements of `structure.identifiers` (i.e. `identifier`) and both are reused below  for determining `subViolations`. It might be worth extracting this into a variable.